ext/fast_xs_extra/fast_xs_extra.java

```diff
@@ -0,0 +1,127 @@
+import org.jruby.Ruby;
+import org.jruby.RubyString;
+import org.jruby.RubyFixnum;
+import org.jruby.anno.JRubyMethod;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.jcodings.Encoding;
+import java.nio.charset.Charset;
+
+public class fast_xs_extra {
+    @JRubyMethod(name = "fast_xs_html")
+    public static IRubyObject fast_xs_html_p(ThreadContext context, IRubyObject self) {
+        RubyString replace;
+        RubyFixnum index;
+        RubyString selfStr = (RubyString) self;
+        String str = selfStr.toString();
+        Ruby runtime = context.runtime;
+        StringBuilder sb = new StringBuilder();
+        Encoding enc = selfStr.getEncoding();
+
+        for (int i = 0; i < str.length(); i++) {
+            char c = str.charAt(i);
+            String entity;
+            if (c == '&') {
+                entity = "&amp;";
+            } else if (c == '<') {
+                entity = "&lt;";
+            } else if (c == '>') {
+                entity = "&gt;";
+            } else if (c == '"') {
+                entity = "&quot;";
+            } else if (c == '\'') {
+                entity = "&#x27;";
+            } else {
+              entity = String.valueOf(c);
+            }
+
+            sb.append(entity);
+        }
+        return RubyString.newString(context.runtime, sb.toString(), enc);
+    }
+
+
+    private static boolean cgiOk(char x) {
+        return ((x >= 'a' && x <= 'z') ||
+          (x >= 'A' && x <= 'Z') ||
+          (x >= '0' && x <= '9') ||
+          (x == '.' || x == '-' || x == '_'));
+    }
+
+    private static IRubyObject fast_xs_uri_encode_p(ThreadContext context, IRubyObject self, boolean escapeSpace) {
+        RubyString replace;
+        RubyFixnum index;
+        RubyString selfStr = (RubyString) self;
+        String str = selfStr.toString();
+        Ruby runtime = context.runtime;
+        StringBuilder sb = new StringBuilder();
+        Encoding enc = selfStr.getEncoding();
+        for (int i = 0; i < str.length(); i++) {
+            char c = str.charAt(i);
+            String entity;
+            if (cgiOk(c)) {
+                entity = String.valueOf(c);
+            } else if (escapeSpace && c == ' ') {
+                entity = "+";
+            } else {
+                RubyFixnum tmpIndex = RubyFixnum.newFixnum(runtime, i);
```

Comment:
I don't like this approach in the else branch at all. @headius maybe you have a better idea?

Reply:
Ok, first off you can ditch all "19" calls (below). They're deprecated in most cases and some are being deleted already on master for 9.2. Just call the bare name equivalent.

Then, for this op_aref call into String, you can either dynamically call it or statically call without using a Fixnum object. Try calling this method on RubyString instead:

```java
    public final IRubyObject substr(Ruby runtime, int beg, int len) {
```